### PR TITLE
Remove wlr_resource_get_buffer_size

### LINKED ISSUE #3015: buffer: get rid of wlr_resource_get_buffer_size
The only reason why we have `wlr_resource_get_buffer_size` is that we historically didn't always have a `wlr_buffer` for all types of `wl_buffer` objects coming from clients. With https://github.com/swaywm/wlroots/pull/2708, this time has passed.

I'd like to remove `wlr_resource_get_buffer_size` because it creates duplication with `wlr_client_buffer_import`: both functions need to handle all existing types of `wl_buffer` objects. This will help with enabling third-party `wl_buffer` implementations as well.

A possible plan is:

- Split `wlr_client_buffer_import` into two functions: one that turns a `wl_resource` into a `wlr_buffer` (e.g. named `wlr_buffer_from_resource`), and another that creates a `wlr_client_buffer` from a `wlr_buffer` (e.g. named `wlr_client_buffer_create`).
- In `wlr_surface_state`, add a new `struct wlr_buffer *buffer` field.
- In `surface_state_finalize`:
  - Before anything else, turn the `buffer_resource` into a `wlr_buffer` (via `wlr_buffer_from_resource`).
  - Get the buffer size from the `wlr_buffer`.
- In `surface_apply_damage`, turn the `wlr_buffer` into a `wlr_client_buffer` (via `wlr_client_buffer_create`), then unlock the `wlr_buffer` (so that the `n_locks` check in `wlr_client_buffer_apply_damage` continues to work).

cc @bl4ckb0ne

### PR DESCRIPTION
**Pull request by bl4ckb0ne**
Fixes #3015 

~~Turns out `wlr_surface_state` buffer member can be avoided, I'm fairly sure there's a way to make it hold only a `wlr_buffer` and get the `buffer_resource`, `buffer_width` and `buffer_height` from it without having to duplicate them. Let me know if you want that in this patch or in another one.~~

cc @emersion 

* * *

Breaking change:

- `wlr_resource_get_buffer_size` has been removed.
- `wlr_client_buffer_import` has been split into two functions: `wlr_buffer_from_resource` and `wlr_client_buffer_create`.

### CODE REVIEW

File: types/wlr_buffer.c [hunk 1]
```diff
@@ -254,7 +228,7 @@ struct wlr_client_buffer *wlr_client_buffer_import(
 	wlr_buffer_lock(&client_buffer->base);
 	wlr_buffer_drop(&client_buffer->base);
 
-	return client_buffer;
+	return &client_buffer->base;
```
Comment (emersion) [inline types/wlr_buffer.c:231]: Returning the client buffer here is not okay. In the case of a `wl_shm` buffer and the GLES2 renderer, the buffer will be uploaded each time `wlr_buffer_import` is called. We should do this a single time per commit. but this patch does it more often since this function is used to get the buffer size.

Comment (emersion) [reply types/wlr_buffer.c:231]: Note, buffer upload == buffer copy == slow :P

Comment (bl4ckb0ne) [reply types/wlr_buffer.c]: I see what I got wrong, made a few changes and the wlr_buffer is imported only once now.

File: types/wlr_buffer.c [hunk 2]
```diff
@@ -226,7 +205,12 @@ struct wlr_client_buffer *wlr_client_buffer_import(
 		wl_resource_post_error(resource, 0, "unknown buffer type");
```
Comment (emersion) [inline types/wlr_buffer.c:205]: The `wlr_log` message above is outdated: this function doesn't upload textures anymore. We can probably just drop the "Cannot upload texture" part.

File: types/wlr_surface.c [hunk 3]
```diff
@@ -169,10 +169,16 @@ static void surface_state_viewport_src_size(struct wlr_surface_state *state,
 
 static void surface_state_finalize(struct wlr_surface *surface,
 		struct wlr_surface_state *state) {
+	if (state->buffer_resource) {
+		state->buffer = wlr_buffer_import(surface->renderer,
+				state->buffer_resource);
+		wlr_buffer_lock(state->buffer);
```
Comment (emersion) [inline types/wlr_surface.c:175]: `wlr_buffer_import` returns a buffer which is already locked. No need to lock it a second time.

File: types/wlr_surface.c [hunk 4]
```diff
@@ -255,6 +261,7 @@ static void surface_state_copy(struct wlr_surface_state *state,
 		struct wlr_surface_state *next) {
 	state->width = next->width;
 	state->height = next->height;
+	state->buffer = next->buffer;
```
Comment (emersion) [inline types/wlr_surface.c:264]: You might notice that `surface_state_copy` is a bit weird: the `buffer_resource` doesn't get copied.

The reason is we have `wlr_surface.previous` which holds the state for the _previous_ commit. We use `surface_state_copy` to construct `previous`. However we don't want to populate the buffer fields of the previous state, because that would lock the buffer for longer than necessary (and force the client to allocate and use one more buffer).

I don't like `wlr_surface.previous`, I think we should get rid of it, but that's another battle (a few more details about this in https://github.com/swaywm/wlroots/issues/2098). In the meantime, move this logic into `surface_state_move` (where you can notice `buffer_resource` is handled as well).

File: types/wlr_surface.c [hunk 5]
```diff
@@ -365,12 +372,13 @@ static void surface_apply_damage(struct wlr_surface *surface) {
...
 
-	struct wlr_client_buffer *buffer =
-		wlr_client_buffer_import(surface->renderer, resource);
+	struct wlr_client_buffer *buffer = wlr_client_buffer_create(
+			surface->current.buffer, surface->renderer, resource);
 	if (buffer == NULL) {
 		wlr_log(WLR_ERROR, "Failed to upload buffer");
 		return;
 	}
+	wlr_buffer_unlock(surface->current.buffer);
```
Comment (emersion) [inline types/wlr_surface.c:381]: We should reset `surface->current.buffer` to NULL here, to avoid having someone read a free'd pointer.

File: types/wlr_buffer.c [hunk 6]
```diff
@@ -194,8 +168,13 @@ static void client_buffer_resource_handle_destroy(struct wl_listener *listener,
 	// which case we'll read garbage. We decide to accept this risk.
 }
 
-struct wlr_client_buffer *wlr_client_buffer_import(
-		struct wlr_renderer *renderer, struct wl_resource *resource) {
+struct wlr_buffer *wlr_buffer_import(struct wlr_renderer *renderer,
+		struct wl_resource *resource) {
+	if (!resource) {
```
Comment (emersion) [inline types/wlr_buffer.c:173]: I think we can just let `wlr_resource_is_buffer` segfault instead. It's a caller error to provide a NULL `wl_resource`.

Segfaults are sometimes better than gracefully handling errors, because they complain loudly and result in user bug reports. Error logs can go unnoticed.

**Review by emersion — CHANGES_REQUESTED** [review]
Thanks, looks much better! Here are a few comments.

File: include/wlr/types/wlr_buffer.h [hunk 7]
```diff
@@ -111,6 +112,14 @@ bool wlr_buffer_get_dmabuf(struct wlr_buffer *buffer,
  */
 bool wlr_buffer_get_shm(struct wlr_buffer *buffer,
 	struct wlr_shm_attributes *attribs);
+/**
+ * Creates a wlr_buffer from a wl_resource and lock it. Once the caller is done
+ * with the buffer, they must call wlr_buffer_unlock.
+ *
+ * The wl_resource must be a wl_buffer.
+ */
+struct wlr_buffer *wlr_buffer_import(struct wlr_renderer *renderer,
```
[the remark this answers is not in the document]
Comment (bl4ckb0ne) [reply include/wlr/types/wlr_buffer.h:121]: `wlr_buffer_from_resource` makes the most sense.

File: include/wlr/types/wlr_buffer.h [hunk 8]
```diff
@@ -111,6 +112,14 @@ bool wlr_buffer_get_dmabuf(struct wlr_buffer *buffer,
  */
 bool wlr_buffer_get_shm(struct wlr_buffer *buffer,
 	struct wlr_shm_attributes *attribs);
+/**
+ * Creates a wlr_buffer from a wl_resource and lock it. Once the caller is done
```
[the remark this answers is not in the document]
Comment (bl4ckb0ne) [reply include/wlr/types/wlr_buffer.h:116]: Is "imports" more suitable?

File: types/wlr_buffer.c (hunk 6 above)
Comment (bl4ckb0ne) [reply types/wlr_buffer.c:173]: I merged it with the other assert to show that we know this is supposed to segfault.

File: include/wlr/types/wlr_buffer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (emersion) [reply include/wlr/types/wlr_buffer.h:116]: Hm, maybe "obtains"?

[the remark this answers is not in the document]
Comment (bl4ckb0ne) [reply include/wlr/types/wlr_buffer.h:116]: "Transforms"?

File: types/wlr_buffer.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (bl4ckb0ne) [reply types/wlr_buffer.c:205]: Inside `wlr_surface_apply_damage`? It kinda makes sense to leave it here for the custom buffers.

File: include/wlr/types/wlr_buffer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (emersion) [reply include/wlr/types/wlr_buffer.h:116]: I'm not a fan of "import" because in the case of DMA-BUFs, this function doesn't import anything. I'm fine with "transform". Other alternatives in the same vein could be "convert" or "turn a wl_resource into a wlr_buffer".

File: types/wlr_buffer.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (emersion) [reply types/wlr_buffer.c:205]: Yup, inside `surface_apply_damage`.

[the remark this answers is not in the document]
Comment (emersion) [reply types/wlr_buffer.c:205]: > It kinda makes sense to leave it here for the custom buffers.

I don't understand, can you elaborate?

[the remark this answers is not in the document]
Comment (bl4ckb0ne) [reply types/wlr_buffer.c:205]: Ah I see, I got confused up. You're saying to put this in `wlr_surface` in case it returns NULL. Nevermind.

File: types/wlr_surface.c (hunk 3 above)
Comment (bl4ckb0ne) [reply types/wlr_surface.c:175]: I'm hitting an assert without it, tracked it down to this one yesterday.

> Assertion failed: buffer->n_locks > 0 (types/wlr_buffer.c: wlr_buffer_unlock: 58)

File: include/wlr/types/wlr_buffer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bl4ckb0ne) [reply include/wlr/types/wlr_buffer.h]: I went with transform.

File: types/wlr_buffer.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (bl4ckb0ne) [reply types/wlr_buffer.c]: I ended up shoving it in `surface_state_finalize` since it's the only place that calls `wlr_buffer_from_resource`.

File: types/wlr_surface.c (hunk 3 above)
Comment (emersion) [reply types/wlr_surface.c:180]: In `surface_state_copy` the previous version was copying the pointer to the `wlr_buffer` without locking (`next->buffer = state->buffer` should have been `next->buffer = wlr_buffer_lock(state->buffer)`). I didn't mention it because I suggested to remove that code.

tl;dr I think it should work better with the new version of the PR. Can you try again removing it?

Comment (bl4ckb0ne) [reply types/wlr_surface.c:180]: Yup, it works

File: types/wlr_buffer.c [hunk 9]
```diff
@@ -219,14 +193,15 @@ struct wlr_client_buffer *wlr_client_buffer_import(
...
 		return NULL;
 	}
+	
+	return buffer;
+}
 
+struct wlr_client_buffer *wlr_client_buffer_create(struct wlr_buffer *buffer,
+		struct wlr_renderer *renderer, struct wl_resource *resource) {
 	struct wlr_texture *texture = wlr_texture_from_buffer(renderer, buffer);
 	wlr_buffer_unlock(buffer);
```
Comment (emersion) [inline types/wlr_buffer.c:206]: This `wlr_buffer_unlock` call makes `wlr_client_buffer_create` a bit weird: it decrements ref'count of the `wlr_buffer` passed in. This is surprising for callers, I think it'd be better to let the callers do it if they need to.

File: types/wlr_surface.c [hunk 10]
```diff
@@ -303,8 +313,12 @@ static void surface_state_move(struct wlr_surface_state *state,
 	surface_state_copy(state, next);
 
 	if (next->committed & WLR_SURFACE_STATE_BUFFER) {
+		if (next->buffer) {
+			state->buffer = wlr_buffer_lock(next->buffer);
+		}
 		surface_state_set_buffer(state, next->buffer_resource);
 		surface_state_reset_buffer(next);
+		next->buffer = NULL;
```
Comment (emersion) [inline types/wlr_surface.c:321]: Here we unset `next->buffer` without decrementing its ref'count.

File: types/wlr_buffer.c (hunk 9 above)
Comment (bl4ckb0ne) [reply types/wlr_buffer.c]: Removed it

File: types/wlr_surface.c (hunk 10 above)
Comment (bl4ckb0ne) [reply types/wlr_surface.c]: Feels a bit weird to lock it when copying it to then unlock it a few lines beneath, but I get why.

Comment (emersion) [reply types/wlr_surface.c]: If you lock it then unlock if immediately, it's possible to simplify to just "transfer" the ref'count.

e.g. it's perfectly fine to change

```c
b = wlr_buffer_lock(a);
wlr_buffer_unlock(a);
a = NULL;
```

to

```c
b = a;
a = NULL;
```

**Comment by emersion** [comment]
I think we should also call `wlr_buffer_unlock` in `surface_state_finish`, to make sure we don't leave buffer refs behind when the surface is destroyed.

File: types/wlr_surface.c [hunk 11]
```diff
@@ -169,10 +169,20 @@ static void surface_state_viewport_src_size(struct wlr_surface_state *state,
 
 static void surface_state_finalize(struct wlr_surface *surface,
 		struct wlr_surface_state *state) {
+	if (state->buffer_resource) {
```
Comment (emersion) [inline types/wlr_surface.c:172]: I think we could move this to the `if ((state->committed & WLR_SURFACE_STATE_BUFFER))` below. No need to update `state->buffer` if it's still the same as before.

Comment (bl4ckb0ne) [reply types/wlr_surface.c]: Done. I think I put it there when I was debugging and forgot to move it after.

File: types/wlr_surface.c [hunk 12]
```diff
@@ -170,9 +170,17 @@ static void surface_state_viewport_src_size(struct wlr_surface_state *state,
 static void surface_state_finalize(struct wlr_surface *surface,
 		struct wlr_surface_state *state) {
 	if ((state->committed & WLR_SURFACE_STATE_BUFFER)) {
-		if (state->buffer_resource != NULL) {
-			wlr_resource_get_buffer_size(state->buffer_resource,
-				&state->buffer_width, &state->buffer_height);
+		if (state->buffer_resource) {
+			state->buffer = wlr_buffer_from_resource(surface->renderer,
```
Comment (emersion) [inline types/wlr_surface.c:174]: When we overwrite a `state->buffer` field, we need to make sure it's NULL. If it's not, it means it's holding a `wlr_buffer` reference, so overwriting it will leak a reference.

File: types/wlr_surface.c [hunk 13]
```diff
@@ -303,9 +311,14 @@ static void surface_state_move(struct wlr_surface_state *state,
 	surface_state_copy(state, next);
 
 	if (next->committed & WLR_SURFACE_STATE_BUFFER) {
+		if (next->buffer) {
+			state->buffer = wlr_buffer_lock(next->buffer);
```
Comment (emersion) [inline types/wlr_surface.c:315]: Same here, if `state->buffer` is not NULL, we must make sure to drop the ref.

**Comment by emersion** [comment]
Addressing the last two comments should make some `wl_shm` clients a bit confused, e.g. GTK clients will die in an assertion failure. This is due to the `wl_buffer_send_release` call in `wlr_client_buffer_apply_damage`. We now sometimes call `wlr_buffer_from_resource` followed by `wlr_client_buffer_apply_damage` followed by `wlr_buffer_unlock`. The last two calls send a `wl_buffer.release` event to the client.

To fix it, we can just remove the `wlr_buffer_from_resource` call from `wlr_client_buffer_apply_damage`, since we now take care of it ourselves.

File: types/wlr_surface.c [hunk 14]
```diff
@@ -365,12 +378,14 @@ static void surface_apply_damage(struct wlr_surface *surface) {
...
-	struct wlr_client_buffer *buffer =
-		wlr_client_buffer_import(surface->renderer, resource);
+	struct wlr_client_buffer *buffer = wlr_client_buffer_create(
+			surface->current.buffer, surface->renderer, resource);
 	if (buffer == NULL) {
 		wlr_log(WLR_ERROR, "Failed to upload buffer");
 		return;
 	}
+	wlr_buffer_unlock(surface->current.buffer);
+	surface->current.buffer = NULL;
```
Comment (emersion) [inline types/wlr_surface.c:388]: Here we take care of unlocking the buffer. But we don't do it in all branches of this function. We should also unlock the buffer above when `return` is used.

File: types/wlr_surface.c [hunk 15]
```diff
@@ -170,9 +170,21 @@ static void surface_state_viewport_src_size(struct wlr_surface_state *state,
 static void surface_state_finalize(struct wlr_surface *surface,
 		struct wlr_surface_state *state) {
 	if ((state->committed & WLR_SURFACE_STATE_BUFFER)) {
-		if (state->buffer_resource != NULL) {
-			wlr_resource_get_buffer_size(state->buffer_resource,
-				&state->buffer_width, &state->buffer_height);
+		if (state->buffer_resource) {
+			if (state->buffer) {
+				wlr_buffer_unlock(state->buffer);
```
Comment (emersion) [inline types/wlr_surface.c:175]: Nit: it's safe to pass NULL to `wlr_buffer_unlock`

**Review by emersion — APPROVED** [review]
LGTM, thanks!
